reactor-netty-http/src/main/java/reactor/netty/http/server/NonSslRedirectDetector.java

```diff
@@ -0,0 +1,130 @@
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslHandler;
+import reactor.netty.NettyPipeline;
+import reactor.netty.tcp.SslProvider;
+
+import javax.annotation.Nullable;
+import java.net.SocketAddress;
+
+import static io.netty.handler.codec.ByteToMessageDecoder.MERGE_CUMULATOR;
+
+/**
+ * @author James Chen
+ */
+final class NonSslRedirectDetector extends ChannelInboundHandlerAdapter {
+
+	private static final int SSL_RECORD_HEADER_LENGTH = 5;
+
+	private final ByteToMessageDecoder.Cumulator cumulator = MERGE_CUMULATOR;
+	private final NonSslRedirectHandler redirectHandler;
+	private final SslProvider sslProvider;
+	private final SocketAddress remoteAddress;
+	private final boolean sslDebug;
+
+	private ByteBuf cumulation;
+
+	public NonSslRedirectDetector(NonSslRedirectHandler redirectHandler, SslProvider sslProvider, @Nullable SocketAddress remoteAddress, boolean sslDebug) {
+		this.redirectHandler = redirectHandler;
+		this.sslProvider = sslProvider;
+		this.remoteAddress = remoteAddress;
+		this.sslDebug = sslDebug;
+	}
+
+	@Override
+	public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+		ByteBuf buf = (ByteBuf) msg;
+		ByteBuf potentialHeader = aggregatePotentialSslRecordHeader(ctx, buf);
+		if (potentialHeader != null) {
+			try {
+				handlePotentialHeader(ctx, potentialHeader);
+			}
+			finally {
+				tryReleaseCumulation();
+			}
+		}
+	}
+
+	@Override
+	public void channelReadComplete(ChannelHandlerContext ctx) {
+		tryReleaseCumulation();
+		ctx.fireChannelReadComplete();
+	}
+
+	@Override
+	public void channelInactive(ChannelHandlerContext ctx) {
+		tryReleaseCumulation();
+		ctx.fireChannelInactive();
+	}
+
+	@Override
+	public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+		if (evt instanceof ChannelInputShutdownEvent) {
+			tryReleaseCumulation();
+		}
+		ctx.fireUserEventTriggered(evt);
+	}
+
+	@Nullable
+	private ByteBuf aggregatePotentialSslRecordHeader(ChannelHandlerContext ctx, ByteBuf buf) {
+		if (cumulation == null) {
+			if (buf.readableBytes() >= SSL_RECORD_HEADER_LENGTH) {
+				return buf;
+			}
+			else {
+				cumulation = Unpooled.EMPTY_BUFFER;
+			}
+		}
+		cumulation = cumulator.cumulate(ctx.alloc(), cumulation, buf);
+		if (cumulation.readableBytes() >= SSL_RECORD_HEADER_LENGTH) {
+			return cumulation;
+		}
+		return null;
+	}
+
+	private void handlePotentialHeader(ChannelHandlerContext ctx, ByteBuf buf) throws Exception {
+		ChannelPipeline pipeline = ctx.pipeline();
+		pipeline.remove(this);
+		if (SslHandler.isEncrypted(buf)) {
+			// Lazy init for better performance
+			sslProvider.addSslHandler(ctx.channel(), remoteAddress, sslDebug);
+			ChannelHandlerContext context = pipeline.context(NettyPipeline.SslHandler);
+			SslHandler sslHandler = (SslHandler) context.handler();
+			sslHandler.channelRead(context, buf);
+		}
+		else {
+			redirectHandler.enableRedirecting();
```

Comment:
Instead of adding `NonSslRedirectHandler` always and then setting a flag here, is it possible to add `NonSslRedirectHandler` here, so the handler will be available in the pipeline only if it is needed.

Reply:
No. If we add `NonSslRedirectHandler` here (earlier), there is no HTTP request aggregator to aggregate and so we cannot get the host from the HTTP request in NonSslRedirectHandler. And that's why I added two handlers.

BTW: NonSslRedirectDetector is added before all HTTP handlers because we need to check whether it's an SSL header or a potential HTTP request.

Reply:
Interesting
I was thinking about something like this

```
		else {
			pipeline.addBefore(NettyPipeline.ReactiveBridge, NettyPipeline.NonSslRedirectHandler, new NonSslRedirectHandler());
			ctx.fireChannelRead(buf);
		}
```

Wdyt?
At that point we will have everything that is needed from HTTP protocol point of view.

Reply:
OK, I will replace with `pipeline.addBefore(NettyPipeline.ReactiveBridge, NettyPipeline.NonSslRedirectHandler, new NonSslRedirectHandler())` and it makes code cleaner.